src/lib/pubkey/sphincsplus/sphincsplus_common/sp_address.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * SPHINCS+ Address
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, René Meusel, Amos Treiber - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ **/
+
+#ifndef BOTAN_SPHINCS_PLUS_ADDRESS_H_
+#define BOTAN_SPHINCS_PLUS_ADDRESS_H_
+
+#include <array>
+
+#include <botan/hash.h>
+#include <botan/internal/loadstor.h>
+#include <botan/internal/sp_types.h>
+
+namespace Botan {
+
+enum class Sphincs_Address_Type : uint32_t {
+   WotsHash = 0,
+   WotsPublicKeyCompression = 1,
+   HashTree = 2,
+   ForsTree = 3,
+   ForsTreeRootsCompression = 4,
+   WotsKeyGeneration = 5,
+   ForsKeyGeneration = 6
+};
+
+class BOTAN_TEST_API Sphincs_Address final {
+   private:
+      static constexpr size_t layer_offset = 0;
+      static constexpr size_t tree_offset = 1;  // 96bits
+      static constexpr size_t type_offset = 4;
+      static constexpr size_t keypair_offset = 5;
+      static constexpr size_t chain_offset = 6;
+      static constexpr size_t hash_offset = 7;
+      static constexpr size_t tree_height_offset = chain_offset;
+      static constexpr size_t tree_index_offset = hash_offset;
+
+   public:
+      using enum Sphincs_Address_Type;
+
+      Sphincs_Address(Sphincs_Address_Type type) {
+         m_address.fill(0);
+         set_type(type);
+      }
+
+      Sphincs_Address(std::array<uint32_t, 8> address) { std::copy(address.begin(), address.end(), m_address.begin()); }
```

Comment:
I think this formatting is not accepted anymore after #3568 - you may need to rebase again and reformat